### PR TITLE
Strip query comments

### LINKED ISSUE #422: question marks in comments are evaluated as query parameters
```php
$query = '
    Select * from ada
    where adakzid = 15 -- customer ???
';

$stmt = $pdo->prepare($query);
$query = $stmt->execute([]);
```

phpstan reports:

```
 ------ ------------------------------------------------------------------------ 
  Line   example.php
 ------ ------------------------------------------------------------------------
     7    Query expects 3 placeholders, but no values are given.
 ------ ------------------------------------------------------------------------

```

### PR DESCRIPTION
**Pull request by marmichalski**
Maybe fixes #422.

Not sure where the `stripComments` method belongs :see_no_evil:

### CODE REVIEW

File: src/Analyzer/QueryPlanQueryResolver.php [hunk 1]
```diff
@@ -37,7 +37,7 @@ public function resolve(Scope $scope, Expr $queryExpr, ?Type $parameterTypes): i
             $normalizedQuery = QuerySimulation::stripTrailers($queryString);
 
             if (null !== $normalizedQuery) {
-                yield $normalizedQuery;
+                yield QuerySimulation::stripComments($normalizedQuery);
```
Comment (marmichalski) [inline src/Analyzer/QueryPlanQueryResolver.php:40]: I dont think this is needed in the query planner..?

File: src/QueryReflection/QueryReflection.php [hunk 2]
```diff
@@ -135,15 +135,15 @@ public function resolveQueryStrings(Expr $queryExpr, Scope $scope): iterable
 
         if ($type instanceof UnionType) {
             foreach (TypeUtils::getConstantStrings($type) as $constantString) {
-                yield $this->normalizeQueryString($constantString->getValue());
+                yield QuerySimulation::stripComments($this->normalizeQueryString($constantString->getValue()));
```
Comment (marmichalski) [inline src/QueryReflection/QueryReflection.php:138]: Might be better to always strip comments as part of `normalizeQueryString`?

File: src/Rules/PdoStatementExecuteMethodRule.php [hunk 3]
```diff
@@ -44,7 +44,7 @@ public function processNode(Node $methodCall, Scope $scope): array
             return [];
         }
 
-        if (PdoStatement::class !== $methodReflection->getDeclaringClass()->getName()) {
+        if (PDOStatement::class !== $methodReflection->getDeclaringClass()->getName()) {
```
Comment (marmichalski) [inline src/Rules/PdoStatementExecuteMethodRule.php:47]: class-name case-mismatch

File: src/Analyzer/QueryPlanQueryResolver.php [hunk 4]
```diff
@@ -34,7 +34,7 @@ public function resolve(Scope $scope, Expr $queryExpr, ?Type $parameterTypes): i
         }
 
         foreach ($queryStrings as $queryString) {
-            $normalizedQuery = QuerySimulation::stripTrailers($queryString);
+            $normalizedQuery = QuerySimulation::stripTrailers(QuerySimulation::stripComments($queryString));
```
Comment (marmichalski) [inline src/Analyzer/QueryPlanQueryResolver.php:37]: guess this order makes more sense

File: src/QueryReflection/QuerySimulation.php [hunk 5]
```diff
@@ -158,4 +157,28 @@ public static function stripTrailers(string $queryString): ?string
...
     }
+
+    /**
+     * @see https://larrysteinle.com/2011/02/09/use-regular-expressions-to-clean-sql-statements/
+     * @see https://github.com/decemberster/sql-strip-comments/blob/3bef3558211a6f6191d2ad0ceb8577eda39dd303/index.js
+     */
+    public static function stripComments(string $query): string
+    {
+        return trim(preg_replace_callback(
+            '/("(""|[^"])*")|(\'(\'\'|[^\'])*\')|(--[^\n\r]*)|(\/\*[\w\W]*?(?=\*\/)\*\/)/m',
```
Comment (akondas) [inline src/QueryReflection/QuerySimulation.php:168]: what kind of magic is this?

File: tests/rules/PdoStatementExecuteMethodRuleTest.php [hunk 6]
```diff
@@ -114,6 +114,28 @@ public function testParameterErrors(): void
...
+            ],
+            [
+                'Query expects placeholder :more, but it is missing from values given.',
+                122,
+            ],
+            [
+                'Query expects placeholder :placeholders, but it is missing from values given.',
+                122,
+            ],
+            */
```
Comment (akondas) [inline tests/rules/PdoStatementExecuteMethodRuleTest.php:138]: why?

**Review by akondas — APPROVED** [review]

File: tests/rules/PdoStatementExecuteMethodRuleTest.php (hunk 6 above)
Comment (marmichalski) [reply tests/rules/PdoStatementExecuteMethodRuleTest.php:138]: Those are the errors reported before implementing fix. They're no longer valid, but leaving them commented as per this files style (I think, there are already commented out ones)

Comment (akondas) [reply tests/rules/PdoStatementExecuteMethodRuleTest.php:138]: ok, make sense :+1:

File: src/QueryReflection/QueryReflection.php [hunk 7]
```diff
@@ -278,10 +278,7 @@ private function resolveConstantArray(ConstantArrayType $parameterTypes, bool $f
         $optionalKeys = $parameterTypes->getOptionalKeys();
 
         foreach ($keyTypes as $i => $keyType) {
-            $isOptional = \in_array($i, $optionalKeys, true);
-            if ($forceOptional) {
-                $isOptional = true;
-            }
+            $isOptional = \in_array($i, $optionalKeys, true) || $forceOptional;
```
Comment (staabm) [inline src/QueryReflection/QueryReflection.php:281]: I don't like this style

File: tests/rules/PdoStatementExecuteMethodRuleTest.php (hunk 6 above)
Comment (staabm) [reply tests/rules/PdoStatementExecuteMethodRuleTest.php:138]: please drop these commented lines. 

the similar stuff in line 18 was commented, as I was not sure about the tests and removing the test-lines would have the consequence that all error expectation had to be adjusted (because removing lines, makes all reported error lines change)

File: src/QueryReflection/QuerySimulation.php [hunk 8]
```diff
@@ -124,14 +124,13 @@ public static function simulateParamValueType(Type $paramType, bool $preparedPar
...
-        $queryString = self::stripTrailers($queryString);
+        $queryString = self::stripTrailers(self::stripComments($queryString));
 
         if (null === $queryString) {
             return null;
         }
-        $queryString .= ' LIMIT 0';
 
-        return $queryString;
+        return $queryString.' LIMIT 0';
```
Comment (staabm) [inline src/QueryReflection/QuerySimulation.php:133]: please drop this change

File: src/QueryReflection/QueryReflection.php [hunk 9]
```diff
@@ -251,7 +251,7 @@ public function resolveParameters(Type $parameterTypes): ?array
 
         if ($parameterTypes instanceof UnionType) {
             foreach (TypeUtils::getConstantArrays($parameterTypes) as $constantArray) {
-                $parameters = $parameters + $this->resolveConstantArray($constantArray, true);
+                $parameters += $this->resolveConstantArray($constantArray, true);
```
Comment (staabm) [inline src/QueryReflection/QueryReflection.php:254]: please revert this change. I don't link `+=` on arrays

File: tests/rules/data/pdo-stmt-execute-error.php [hunk 10]
```diff
@@ -98,4 +98,27 @@ public function bug336Named(PDO $pdo)
...
+        $stmt->execute([]);
+    }
+
+    public function bug422NamedPlaceholdersInComment(PDO $pdo)
+    {
+        $stmt = $pdo->prepare('SELECT email, adaid FROM ada -- nice :placeholder bro');
+        $stmt->execute([]);
+    }
+
+    public function bug422NamedAndUnnamedPlaceholdersInComment(PDO $pdo)
```
Comment (staabm) [inline tests/rules/data/pdo-stmt-execute-error.php:119]: please add a additional est which has parameters and a comment in the query

**Review by staabm — CHANGES_REQUESTED** [review]
this looks awesome - thank you

File: tests/rules/data/pdo-stmt-execute-error.php (hunk 10 above)
Comment (marmichalski) [reply tests/rules/data/pdo-stmt-execute-error.php:119]: :gift_heart:

**Comment by staabm** [comment]
Thank you 🙏
